xde/src/dls/sys.rs

```diff
@@ -0,0 +1,189 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2024 Oxide Computer Company
+
+// stuff we need from dls
+
+use crate::ip::kcondvar_t;
+use crate::ip::krwlock_t;
+use crate::ip::list_node_t;
+use crate::ip::major_t;
+use crate::ip::minor_t;
+use crate::ip::queue_t;
+use crate::ip::t_uscalar_t;
+use crate::mac;
+use crate::mac::mac_client_handle;
+use crate::mac::mac_handle;
+use crate::mac::mac_promisc_handle;
+use crate::mac::mac_tx_cookie_t;
+use core::ffi::c_void;
+use illumos_sys_hdrs::boolean_t;
+use illumos_sys_hdrs::c_char;
+use illumos_sys_hdrs::c_int;
+use illumos_sys_hdrs::c_uint;
+use illumos_sys_hdrs::datalink_id_t;
+use illumos_sys_hdrs::kmutex_t;
+use illumos_sys_hdrs::mblk_t;
+use illumos_sys_hdrs::uintptr_t;
+use illumos_sys_hdrs::zoneid_t;
+
+extern "C" {
+    pub fn dls_devnet_create(
+        mh: *mut mac::mac_handle,
+        linkid: datalink_id_t,
+        zoneid: zoneid_t,
+    ) -> c_int;
+
+    pub fn dls_devnet_destroy(
+        mh: *mut mac::mac_handle,
+        linkid: *mut datalink_id_t,
+        wait: boolean_t,
+    ) -> c_int;
+
+    pub fn dls_mgmt_get_linkid(
+        name: *const c_char,
+        linkid: *mut datalink_id_t,
+    ) -> c_int;
+}
+
+// Private DLS functions needed to have a Tx path on top of
+// an existing link while circumventing `ip`.
+extern "C" {
+    pub type dls_devnet_s;
+    pub type dls_link;
+
+    /// Transmit a packet chain on a given link.
+    /// This is effectively one layer above mac_tx.
+    pub fn str_mdata_fastpath_put(
+        dsp: *const dld_str_s,
+        mp: *mut mblk_t,
+        f_hint: uintptr_t,
+        flag: u16,
+    ) -> mac_tx_cookie_t;
+
+    // NOTE: ALL BELOW FUNCTIONS REQUIRE THE MAC PERIMETER TO BE HELD.
+    pub fn dls_devnet_hold(
+        link: datalink_id_t,
+        ddhp: *mut dls_dl_handle,
+    ) -> c_int;
+
+    pub fn dls_devnet_rele(dlh: dls_dl_handle);
+
+    pub fn dls_link_hold(
+        name: *const c_char,
+        dlpp: *mut *mut dls_link,
+    ) -> c_int;
+
+    pub fn dls_link_rele(dlp: *mut dls_link);
+
+    pub fn dls_devnet_mac(dlh: dls_dl_handle) -> *const c_char;
+
+    pub fn dls_open(
+        dlp: *mut dls_link,
+        ddh: dls_dl_handle,
+        dsp: *mut dld_str_s,
+    ) -> c_int;
+
+    pub fn dls_close(dsp: *mut dld_str_s);
+}
+
+#[repr(C)]
+pub enum dld_str_mode_t {
+    DLD_UNITDATA,
+    DLD_FASTPATH,
+    DLD_RAW,
+}
+
+#[repr(C)]
+pub enum dld_passivestate_t {
+    DLD_UNINITIALIZED,
+    DLD_PASSIVE,
+    DLD_ACTIVE,
+}
+
+// Direct translation of the illumos type type, with some
+// pointer fields left as opaque void*s for simplicity.
+#[repr(C)]
+pub struct dld_str_s {
```

Comment:
I think `dld_str_t` is just an opaque struct in `dls.h`, since they're usually allocated/constructed when coming in through `dld_open`. I'm not quite sure if there are other consequences for spinning up a full STREAMS session on the link though.